File: mcrpc.py
import collections
import glob
import hashlib
import json
import os
import re
import shutil
import sys
from distutils.dir_util import copy_tree
from pprint import pprint
from zipfile import ZipFile

import requests
from platformdirs import user_cache_dir

PLATFORM = sys.platform
if PLATFORM.startswith('linux'):
    PATH_DOTMC = os.path.join(os.path.expanduser('~'), '.minecraft')
elif PLATFORM.startswith('darwin'):
    PATH_DOTMC = os.path.join(os.path.expanduser('~'), 'Library', 'Application Support', 'minecraft')
elif PLATFORM.startswith('win32'):
    PATH_DOTMC = os.path.join(os.path.expandvars(r'%APPDATA%'), '.minecraft')

APPNAME = "mcrpc"
PATH_CACHE = user_cache_dir(APPNAME)
PATH_CACHE_JARS = os.path.join(PATH_CACHE, 'jars')
PATH_RP_DIR = os.path.join(PATH_DOTMC, 'resourcepacks')
PATH_TARGET = os.path.join(PATH_CACHE, 'target')
PATH_DEFAULT = os.path.join(PATH_CACHE, 'default')
PATH_CLIENT_LOW = os.path.join(PATH_CACHE, 'client_low.jar')
PATH_CLIENT_HIGH = os.path.join(PATH_CACHE, 'client_high.jar')
LINK_VM = 'https://launchermeta.mojang.com/mc/game/version_manifest_v2.json'
BUF_SIZE = 65536
SESSION = requests.Session()


def choose_rp():
    list_rp = os.listdir(PATH_RP_DIR)
    list_rp_checked = []

    list_rp.sort()

    print('\nSelect the Resource Pack you want to compare:')

    for item in list_rp:
        if item[-4:] == '.zip':
            with ZipFile(os.path.join(PATH_RP_DIR, item), 'r') as zip1:
                if 'pack.mcmeta' in zip1.namelist():
                    list_rp_checked.append(item)
        elif os.path.isdir(os.path.join(PATH_RP_DIR, item)):
            if os.path.isfile(os.path.join(PATH_RP_DIR, item, 'pack.mcmeta')):
                list_rp_checked.append(item)

    if len(list_rp_checked) == 0:
        print('No valid resource packs. Install one and try again.')
        exit(1)

    for i, item in enumerate(list_rp_checked, 1):
        print("  [{}] {}".format(i, item))
    print('Resource Pack choice: ', end='')

    choice_rp_num = input()
    choice_rp = None

    for i, item in enumerate(list_rp_checked, 1):
        if i == int(choice_rp_num):
            choice_rp = item

    return(choice_rp)


def choose_mcv(low=None):
    json_vm = SESSION.get(LINK_VM)
    json_vm.raise_for_status()
    json_vm = json_vm.json()

    print('\nSee non-stable versions? (y/n)')
    choice_stable = input()

    if low == True:
        print('\nSelect the lower Minecraft Version:')
    elif low == False:
        print('\nSelect the higher Minecraft Version:')
    else:
        print('\nSelect the Minecraft Version you want to compare against:')

    i = 1
    counter_size = 0
    choice_mcv_num = ''
    skip_display = False
    while True:
        for item in json_vm['versions']:

            counter_size += 1
            if counter_size > len(json_vm['versions']):
                skip_display = True

            if choice_stable == 'n' or choice_stable == 'N':
                if not item['type'] == 'release':
                    continue

            if not skip_display:
                print("  [{}] {}".format(i, item['id']))

            if i % 25 == 0:
                print('(Hit Enter to show more versions) Minecraft Version choice: ', end='')

                choice_mcv_num = input()
                if choice_mcv_num.isdigit():
                    break
            i += 1
        if choice_mcv_num.isdigit():
            break

    choice_mcv = None

    i = 1
    for item in json_vm['versions']:
        if choice_stable == 'n' or choice_stable == 'N':
            if not item['type'] == 'release':
                continue
        if i == int(choice_mcv_num):
            choice_mcv = item
        i += 1

    return(choice_mcv)


def setup_compare_folders(rp, mcv):
    if os.path.exists(PATH_TARGET):
        shutil.rmtree(PATH_TARGET)
    if os.path.exists(PATH_DEFAULT):
        shutil.rmtree(PATH_DEFAULT)

    if not isinstance(rp, collections.abc.Mapping):
        path_rp = os.path.join(PATH_RP_DIR, rp)
        if path_rp[-4:] == ".zip":
            with ZipFile(path_rp) as zip1:
                for file in zip1.namelist():
                    if file.startswith('assets/'):
                        zip1.extract(file, path=PATH_TARGET)
        else:
            copy_tree(os.path.join(path_rp, 'assets'), os.path.join(PATH_TARGET, 'assets'))
    else:
        path_client = os.path.join(PATH_CACHE_JARS, '{}.jar'.format(rp['id']))
        with ZipFile(path_client) as zip2:
            for file in zip2.namelist():
                if file.startswith('assets/'):
                    zip2.extract(file, path=PATH_TARGET)

    path_client = os.path.join(PATH_CACHE_JARS, '{}.jar'.format(mcv['id']))
    with ZipFile(path_client) as zip3:
        for file in zip3.namelist():
            if file.startswith('assets/'):
                zip3.extract(file, path=PATH_DEFAULT)


def get_client(mcv):
    json_version = SESSION.get(mcv['url'])
    json_version.raise_for_status()
    json_version = json_version.json()

    link_client = json_version['downloads']['client']['url']
    sha_client = json_version['downloads']['client']['sha1']
    sha1 = hashlib.sha1()
    sha_calculated = None

    if not os.path.exists(PATH_CACHE_JARS):
        os.makedirs(PATH_CACHE_JARS)

    path_client = os.path.join(PATH_CACHE_JARS, '{}.jar'.format(mcv['id']))

    if os.path.exists(path_client):
        with open(path_client, 'rb') as f:
            while True:
                data = f.read(BUF_SIZE)
                if not data:
                    break
                sha1.update(data)
        sha_calculated = sha1.hexdigest()

    if not sha_calculated == sha_client:
        with open(path_client, 'wb') as f:
            print('Downloading client.jar…')
            download = SESSION.get(link_client)
            f.write(download.content)


def compare(rp, mcv):
    if isinstance(rp, collections.abc.Mapping):
        # Comparing 2 minecraft versions
        get_client(rp)
        get_client(mcv)
    else:
        # Comparing a resource pack against a minecraft version
        get_client(mcv)

    setup_compare_folders(rp, mcv)

    dir_left = os.path.join(PATH_DEFAULT, '**', '*.*')
    dir_right = os.path.join(PATH_TARGET, '**', '*.*')

    files_left = glob.glob("{}".format(dir_left), recursive=True)
    files_right = glob.glob("{}".format(dir_right), recursive=True)

    if PLATFORM.startswith('linux') or PLATFORM.startswith('darwin'):
        pattern = re.compile(".*/assets/(.*)")
    elif PLATFORM.startswith('win32'):
        pattern = re.compile(".*\\assets\\(.*)")

    for i, file in enumerate(files_left):
        files_left[i] = pattern.sub("\g<1>", file)
    for i, file in enumerate(files_right):
        files_right[i] = pattern.sub("\g<1>", file)

    missing_files = [x for x in files_left if x not in files_right]

    print('\nMissing files:')
    for file in sorted(missing_files):
        print("  {}".format(file))
    print('\nTotal missing files:', len(missing_files))


if __name__ == '__main__':

    while True:
        print('\n[1] Compare a resource pack against a minecraft version')
        print('[2] Compare a minecraft version against another minecraft version')
        print('Choice: ', end='')

        choice_logic = input()

        if choice_logic == '1':
            compare(choose_rp(), choose_mcv())
            break
        elif choice_logic == '2':
            compare(choose_mcv(low=True), choose_mcv(low=False))
            break
        else:
            print('\nInvalid choice!')

    print('\nExit? (y/N)')
    while True:
        choice_quit = input()
        if choice_quit == 'y' or choice_quit == 'Y':
            exit(0)

File: mcrpc.cpp
#include "mcrpc.h"

#include <QCoreApplication>
#include <QCryptographicHash>
#include <QEventLoop>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QNetworkReply>
#include <QNetworkRequest>
#include <QStandardPaths>

#include <filesystem>

#include "JlCompress.h"

mcrpc::mcrpc(QObject *parent)
    : QObject{parent}, qout(stdout), qin(stdin)
{
    // Set .minecraft directory
    dirDotMC = dirHome;
    // Consider using #if defined(Q_OS_LINUX)
    if (platform == "linux")
    {
        if (!dirDotMC.cd(".minecraft"))
        {
            qFatal(".minecraft does not exist!");
        }
    }
    else if (platform == "darwin")
    {
        if (!dirDotMC.cd("Library/Application Support/minecraft"))
        {
            qFatal("~/Library/Application Support/minecraft does not exist!");
        }
    }
    else if (platform == "winnt")
    {
        if (!dirDotMC.cd("AppData/.minecraft"))
        {
            qFatal(".minecraft does not exist!");
        }
    }

    // TODO(txtsd): Use mkdir instead of mkpath
    // Set cache directory
    QDir dirCache(QStandardPaths::writableLocation(QStandardPaths::CacheLocation));
    if (!dirCache.exists())
    {
        dirCache.mkpath(dirCache.absolutePath());
    }

    // Set jars directory
    if (!dirCache.exists("jars"))
    {
        dirCache.mkpath(dirCache.absolutePath() + "/jars");
    }
    dirJars = dirCache;
    dirJars.cd("jars");

    // Set resource packs directory
    if (!dirDotMC.exists("resourcepacks"))
    {
        dirDotMC.mkpath(dirDotMC.absolutePath() + "/resourcepacks");
    }
    dirRP = dirDotMC;
    dirRP.cd("resourcepacks");
}

void mcrpc::run()
{
    while(true)
    {
        qout << "\n" << "[1] Compare a resource pack against a minecraft version" << "\n";
        qout << "[2] Compare a minecraft version against another minecraft version" << "\n";
        qout << "Choice: ";
        qout.flush();

        QString choiceLogic = qin.readLine();

        if (choiceLogic == '1')
        {
            rp = chooseRP();
            mcv1 = chooseMCV();
            compare(rp, mcv1, mcv2);
            break;
        }
        else if (choiceLogic == '2')
        {
            mcv1 = chooseMCV();
            mcv2 = chooseMCV();
            compare(rp, mcv1, mcv2);
            break;
        }
        else
        {
            qout << "\n" << "Invalid choice!" << "\n";
            qout.flush();
        }
    }

    // Quit after everything is done
    // QCoreApplication::quit();
}

QFileInfo mcrpc::chooseRP()
{
    QFileInfoList listRP = dirRP.entryInfoList(QDir::AllDirs | QDir::Files | QDir::NoDotAndDotDot);

    QFileInfoList listRPChecked;

    for(const QFileInfo &item : listRP)
    {
        if (item.suffix() == "zip")
        {
            QStringList archiveContents = JlCompress::getFileList(item.absoluteFilePath());
            if (archiveContents.contains("pack.mcmeta"))
            {
                listRPChecked.append(item);
            }
        }
        else if (QFileInfo::exists(item.absoluteFilePath() + "/pack.mcmeta"))
        {
            listRPChecked.append(item);
        }
    }

    if (listRPChecked.isEmpty())
    {
        qout << "\n" << "No valid resource packs. Install one and try again.";
        qout.flush();
        QCoreApplication::quit();
    }

    while(true)
    {
        qout << "\n" << "Select the Resource Pack you want to compare:";
        qout.flush();

        int i = 1;
        for(const QFileInfo &item : listRPChecked)
        {
            qout << "\n " << "[" << i << "] " << item.fileName();
            qout.flush();
            i++;
        }
        qout << "\n" << "Resource Pack choice: ";
        qout.flush();

        QString choiceRPNum = qin.readLine();

        i = 1;
        for(const QFileInfo &item : listRPChecked)
        {
            if (QString::number(i) == choiceRPNum)
            {
                return item;
            }
            i++;
        }
        qout << "\n" << "Invalid choice!" << "\n";
    }

}

QJsonObject mcrpc::chooseMCV()
{
    QEventLoop loop;
    QObject::connect(&qnam, &QNetworkAccessManager::finished,
        &loop, &QEventLoop::quit);
    QNetworkReply *reply = qnam.get(QNetworkRequest(linkVM));
    loop.exec();

    QJsonObject jsonObject;

    QUrl url = reply->url();
    if (reply->error() == QNetworkReply::NoError)
    {
        QByteArray response(reply->readAll());
        QJsonDocument jsonResponse = QJsonDocument::fromJson(response);
        jsonObject = jsonResponse.object();

    }
    reply->deleteLater();

    QString choiceStable;
    qout << "\n";
    qout.flush();
    while(true)
    {
        qout << "See non-stable versions? (y/N)" << "\n";
        qout.flush();
        choiceStable = qin.readLine();
        if (choiceStable == "y" || choiceStable == "Y" || choiceStable == "n" || choiceStable == "N")
        {

            break;
        }
    }

    qout << "\n" << "Select the Minecraft Version you want to compare against:" << "\n";
    qout.flush();

    QString choiceMCVNum;
    int counter = 0;
    int counterActual = 1;
    int listSize = 25;
    bool skipDisplay = false;
    while(true)
    {
        bool ok = false;
        for(QJsonValue entry : jsonObject["versions"].toArray())
        {
            counter++;

            if (counter >= jsonObject["versions"].toArray().size())
            {
                skipDisplay = true;
            }

            if (choiceStable == "n" || choiceStable == "N")
            {
                if (entry.toObject()["type"].toString() != "release")
                {
                    continue;
                }
            }

            if (!skipDisplay)
            {
                qout << "  [" << counterActual << "] " << entry.toObject()["id"].toString() << "\n";
                qout.flush();
            }

            if (counterActual % listSize == 0)
            {
                qout << "(Hit Enter to show more versions) Minecraft Version choice: ";
                qout.flush();

                choiceMCVNum = qin.readLine();
                // bool ok;
                choiceMCVNum.toInt(&ok);
                if (ok)
                {
                    break;
                }
            }

            counterActual++;
        }
        if (ok)
        {
            break;
        }
    }

    QJsonObject choiceMCV;
    counterActual = 1;
    for(QJsonValue entry : jsonObject["versions"].toArray())
    {
        if (choiceStable == "n" || choiceStable == "N")
        {
            if (entry.toObject()["type"].toString() != "release")
            {
                continue;
            }
        }

        if (QString::number(counterActual) == choiceMCVNum)
        {
            choiceMCV = entry.toObject();
        }

        counterActual++;
    }

    return choiceMCV;

}

void mcrpc::compare(const QFileInfo &rp, const QJsonObject &mcv1, const QJsonObject &mcv2)
{
    if (rp.exists() && !mcv1.isEmpty() && mcv2.isEmpty())
    {
        getClient(mcv1);
        // setupCompareFolders(true);
        QStringList rpList;
        if (rp.isFile())
        {
            rpList = JlCompress::getFileList(rp.absoluteFilePath());
        }
        else
        {
            QDir selectRPDir = rp.dir();
            selectRPDir.cd(rp.fileName());

            // this is not recursive
            // rpList = selectRPDir.entryList(QDir::AllDirs | QDir::Files | QDir::NoDotAndDotDot);

            // actually get all files
            for (std::filesystem::recursive_directory_iterator i(selectRPDir.path().toStdString()), end; i != end; ++i)
            {
                QString entry = QString::fromStdString(i->path().string());
                entry.replace(selectRPDir.path() + QDir::separator(), "");
                rpList.append(entry);
            }
        }
        // qDebug() << rpList << "\n" << rpList.size();
        qDebug() << rpList.size();
        QString clientPathStr = dirJars.absoluteFilePath(mcv1["id"].toString() + ".jar");
        QStringList mcv1List = JlCompress::getFileList(clientPathStr);
        QStringList mcv1ListFiltered;
        for (const QString &item : mcv1List)
        {
            if (item.indexOf("assets") == 0)
            {
                mcv1ListFiltered.append(item);
            }
        }
        qDebug() << mcv1ListFiltered.size();
        // Now compare lists and list missing files
        QStringList results;
        for (const QString &item : mcv1ListFiltered)
        {
            if (!rpList.contains(item, Qt::CaseInsensitive))
            {
                results.append(item);
            }
        }
        // qDebug() << "\n" << "Missing Files: " << "\n" << results << "\n" << results.size();
        qDebug() << results.size();
    }
    else if (!rp.exists() && !mcv1.isEmpty() && !mcv2.isEmpty())
    {
        getClient(mcv1);
        getClient(mcv2);
        // TODO(txtsd): Add logic to always compare older against newer
        // setupCompareFolders();
        QString clientPathStr1 = dirJars.absoluteFilePath(mcv1["id"].toString() + ".jar");
        QStringList mcv1List = JlCompress::getFileList(clientPathStr1);
        QStringList mcv1ListFiltered;
        for (const QString &item : mcv1List)
        {
            if (item.indexOf("assets") == 0)
            {
                mcv1ListFiltered.append(item);
            }
        }
        qDebug() << mcv1ListFiltered.size();
        QString clientPathStr2 = dirJars.absoluteFilePath(mcv2["id"].toString() + ".jar");
        QStringList mcv2List = JlCompress::getFileList(clientPathStr2);
        QStringList mcv2ListFiltered;
        for (const QString &item : mcv2List)
        {
            if (item.indexOf("assets") == 0)
            {
                mcv2ListFiltered.append(item);
            }
        }
        qDebug() << mcv2ListFiltered.size();
        QStringList results;
        for (const QString &item : mcv2ListFiltered)
        {
            if (!mcv1ListFiltered.contains(item, Qt::CaseInsensitive))
            {
                results.append(item);
            }
        }
        // qDebug() << "\n" << "Missing Files: " << "\n" << results << "\n" << results.size();
        qDebug() << results.size();
    }
}

void mcrpc::getClient(const QJsonObject &mcv)
{
    QEventLoop loop;
    QObject::connect(&qnam, &QNetworkAccessManager::finished, &loop, &QEventLoop::quit);
    QNetworkReply *reply = qnam.get(QNetworkRequest(mcv["url"].toString()));
    loop.exec();

    QJsonObject jsonObject;

    QUrl url = reply->url();
    qDebug() << "url" << url;
    if (reply->error() == QNetworkReply::NoError)
    {
        QByteArray response(reply->readAll());
        QJsonDocument jsonResponse = QJsonDocument::fromJson(response);
        jsonObject = jsonResponse.object();
    }
    else
    {
        qCritical() << "1 Failed to download" << mcv["url"].toString();
    }
    reply->deleteLater();

    // QJsonObject _temp1 = jsonObject["downloads"].toObject();
    // qDebug() << "_temp1" << _temp1;

    QString clientLink = jsonObject["downloads"].toObject()["client"].toObject()["url"].toString();
    QString clientSha = jsonObject["downloads"].toObject()["client"].toObject()["sha1"].toString();

    // Check if these turn up empty
    qDebug() << "clientLink" << clientLink;
    qDebug() << "clientSha" << clientSha;

    QString clientPathStr = dirJars.absoluteFilePath(mcv["id"].toString() + ".jar");
    QFile clientPath(clientPathStr);
    // qDebug() << clientPath;

    QCryptographicHash cryptoHash(QCryptographicHash::Sha1);
    QByteArray hashByteArray = "";
    clientPath.open(QIODevice::ReadOnly);

    if (clientPath.exists())
    {
        qDebug() << "Client" << mcv["id"].toString() << "exists";
        // check and set sha1
        while(!clientPath.atEnd())
        {
            cryptoHash.addData(clientPath.readLine());
        }
        clientPath.close();
        hashByteArray = cryptoHash.result();
    }

    qDebug() << hashByteArray.toHex();

    // if sha1 does not match
    // download client
    if (QString(hashByteArray.toHex()) != clientSha)
    {
        qDebug() << "Downloading" << mcv["id"].toString();
        QEventLoop loop2;
        QNetworkReply *reply2 = qnam.get(QNetworkRequest(clientLink));
        QObject::connect(&qnam, &QNetworkAccessManager::finished,
            &loop2, &QEventLoop::quit);
        QObject::connect(reply2, &QNetworkReply::readyRead,
            [this, &clientPath, &reply2](){
            clientPath.write(reply2->readAll());
        });
        clientPath.open(QIODevice::WriteOnly | QIODevice::Append);
        clientPath.resize(0);
        loop2.exec();
        clientPath.close();
        reply2->deleteLater();
    }

    // TODO(txtsd): Check checksum after downloading
}
